sherpa/utils/__init__.py

```diff
@@ -1452,51 +1457,144 @@ def create_expr(vals, mask=None, format='%s', delim='-'):
         index = numpy.arange(len(mask))
         seq = index[mask]
 
-    # diffs has 1 less element than vals
-    #
-    diffs = numpy.apply_along_axis(numpy.diff, 0, seq)
-    diffs = diffs == 1
-
-    def filt(start, end):
-        "What is the filter expression for this range?"
-        vstr = format % start
-        if start != end:
-            vstr += delim + format % end
-        return vstr
-
-    # The trick here is that if diffs is True then we just want to
-    # update the end counter, which gives the end-point of the current
-    # range, and move to the next bin. We only output data when diffs
-    # is False, when we report the previous range and start a new
-    # range.
-    #
-    expr = []
+    exprs = []
     start = vals[0]
-    end = vals[0]
-    for delta, val in zip(diffs, vals[1:]):
 
-        # If this is part of a contiguous sequence then we update
-        # the end value and move on.
-        #
-        if delta:
-            end = val
-            continue
+    # We follow create_expr_int but instead of having separate lo/hi
+    # always we use the same array
+    #
+    startbins = vals[1:]
+    endbins = vals[:-1]
 
-        # If we have a break then output the range we have just
-        # processed.
-        #
-        expr.append(filt(start, end))
+    diffs = numpy.diff(seq)
+    idxs, = numpy.where(diffs != 1)
+    for idx in idxs:
+        exprs.append((start, endbins[idx]))
+        start = startbins[idx]
 
-        # Start a new range.
-        #
-        start = val
-        end = val
+    exprs.append((start, vals[-1]))
+
+    def filt(lo, hi):
+        vstr = format % lo
+        if lo == hi:
+            return vstr
+
+        return vstr + f"{delim}{format % hi}"
+
+    return ",".join([filt(*expr) for expr in exprs])
+
+
+def create_expr_int(lovals, hivals, mask=None,
+                    format='%s', delim='-',
+                    eps=numpy.finfo(numpy.float32).eps):
+    """Create a string representation of a filter (integrated)
+
+    Use the mask to convert the input values into a set of
+    comma-separated filters - low value and high value, separated by
+    the delimeter - that represent the data. Unlike create_expr this
+    routines uses the lovals values for the start of the bin and
+    hivals for the end of each bin, and assumes that contiguous bins
+    should be combined.
+
+    Parameters
+    ----------
+    lovals, hivals : sequence
+        The lower and upper values of each bin. It is required that
+        they are in ascending order and locals < hivals.
+    mask : sequence of bool or None, optional
+        The mask setting for the full dataset, without any filtering
+        applied. A value of True indicates the element is included
+        and False means it is excluded.
```

Comment:
Wait, I don' understand how mask can have a different number of entries than lovals and higvals (See examples below). When I hear "mask" I think "masked bins". In
```
>>> create_expr_int([0.1, 0.2, 0.6, 0.8], [0.2, 0.4, 0.8, 1.0], [True, True, False, True, True])
```
I have four bins, but five mask values. What does that fifth value correspond to? Maybe I've used numpy too much to understand this concept here.
I don't understand that from the docs on this line and I don't understand it from the examples. Instead of "mask", this is more like a "grouping_flag", where True means "start new bin or previous bin continues" and False means "bins ends or if previous was false, skip this bin". If my interpretation is correct, I suggest to rename the parameter and reword the docstring.

After writing this long text, I will now go an read the code to try to understand from there what exactly a mask means here.

Reply:
This is a very-confusing part of the code (and is similar to `create_expr` but I think there are subtle differences). I am trying to follow the existing code.

The lo/hi values are the selected (ie filtered) range (potentially after grouping but one of the changes here is that it doesn't matter in the end) and the mask is for all values, so mask is going to be longer if not(mak.all()) - ie there's a False in there.

The issue is that with lo=[0.1, 0.2, 0.4, 0.5], hi=[0.2, 0.3, 0.45, 0.6] should this be mapped to

- `0.1-0.6`
- `0.1-0.3,0.4-0.45,0.5-0.6` 
- `0.1-0.45,0.5-0.6`
- `0.1-0.3,0.4-0.6`

That is, is the lo,hi grid you sent, which is not continuous, technically continuous or is it missing bins. We use the mask field to indicate where the "edges" are. So

mask = [True] * 4 would generate '0.1-0.6'
mask = [True, True, False, True, False, True] would generate`0.1-0.45,0.5-0.6`
mask = [True, True, True, False, True] generates `0.1-0.45,0.5-0.6`
mask = [True, True, False, True, True] generates `0.1-0.3,0.4-0.6`

We could change how this is done (ie send in the whole lo/hi grid, not just the filtered grid) but that would be a larger change.

I disagree with renaming it, as it is the mask you are thinking of (and we've had the parameter labelled `mask` in `create_expr`).

Reply:
My problem is that I don't understand how the mask grid maps to the edges if there are more mask entries than bins. For example, 
mask = [True, True, False, True, False, True] would generate0.1-0.45,0.5-0.6

I have three groups of True, but the output has only two range. Why? 
The next example
mask = [True, True, True, False, True] generates 0.1-0.45,0.5-0.6

has two groups of True, and also two ranges. So apparently the "number of True groups" is not correlated with "number of ranges listed in output". At the same time, the "Number of elements in mask" can differ from "number of bins". Apparently, I still don't understand. If you can explain it, that would be great, if not, then we can decide that this is not meant to be understood except by the developers, who can read the code and figure it out.

Yet another option is to change it to an easier to understand scheme, say, send in the whole lo/hi grid) and open a new isse and leave fixing up the docs to that issue. (Which you would have to address, since I don't understand it).

Reply:
The number of True elements in the mask matches the size of the lo/hi arrays. The lo/hi arrays are the "masked" data and so do not include bins that have been masked out. We can use the mask array (which is the mask of the actual Data object) to identify which sets of lo/hi bins are "adjacent" [we can not assume a check like hi_i == lo_i+1 as we do not require the bins to be contiguous is general, and even more so for Data1D cases where there is no a-priori relationship between x_i and x_i+1].

If we have a `Data1DInt` case which has

| mask |   T  |   T  |   F  |   T  |  F  |  T  |
| --- | --- | --- |  --- | --- | --- | --- |
| xlo     | 0.1 | 0.2 | 0.3 | 0.4 | 0.5 | 0.6 |
| xhi     | 0.2 | 0.3 | 0.4 | 0.5 | 0.6 | 0.7 |

that is we've ignored the 0.3-0.4 and 0.5-0.6 groups then we send in

    lo = [0.1, 0.2, 0.4, 0.6]
    hi = [0.2, 0.3, 0.5, 0.7]
    mask = [T, T, F, T, F, T ]

to `create_expr_integrated` then know that (0.1,0.2) and (0.2, 0.3) are from the first group of T's, (0.4,0.5) is the second T, and (0.6-0.7) is the last T, which means we can create a filter expression: 0.1-0.3,0.4-0.5,0.6-0.7

There is a similar scheme in `create_expr`.

This is not new (i.e. it is how the code used to work), I've just split the integrated and non-integrated code paths (and done some other shenanigans). I am not going to change this scheme at this time as I don't have time and it is trickier than you might think when you originally decided to fix this bug. It may well be that we can make the code more obvious, but I see that as being done at a later date.

Reply:
I still don't understand, but I agree that we an leave that for a later date.